### PR TITLE
Service runtime V2

### PR DESCRIPTION
**Pull request by aleksmaus**
## What does this PR do?

* Implements service runtime component for V2.
* Extends endpoint spec with some additional attributes for service start/stop/status checks and creds discovery. The creds discovery logic is taken from V1, cleaned up and extracted into its own file, added utz.
* Implements service uninstall
* Refactors ```pkg/core/process/process.go``` adds additional options that are needed for the ```service_command``` implementation.
* Changes ```ComponentsModifier``` to access raw config, needed for the EndpointComponentModifier
* Injects ```host.id``` into configuration, needed for Endpoint
* Injects ```fleet``` and ```policy.revision``` configuration into the Endpoint input configuration
* Bumps the version to 8.6.0 to make it consistent with current beats V2 branch
* Addresses linter complains on affected files

Tested on linux (Ubuntu) and windows(10) with the Endpoint end to end. Confirmed can start stop service and get V2 check-ins. Could not test with unsigned Endpoint on macOS even after following on recommendation for disabling the signature checks etc.

Had issue on with windows Endpoint not sending STOPPED check-in which causes the agent to sit blocked waiting on this until until the stop times out during shutdown (as expected). Following through on this over the next week. Wanted to open this as PR for review at this point.
Should we set the runtime to the failed state on stop if this happens or log and set it to the stopped state? Setting it to the failed state prevents the configuration change application that require restart.
The weak point here is if something goes wrong with gRPC the agent will gets stuck in the failed component state on windows (since we can't use the service manager to stop the service), while on linux  and macOS we can fallback on stopping via systemd/launchd/etc. Will think what's better next week. 
UPDATE: the experience is not great if "service runtime" is left stop in the failed state. Thinking that after attempting every venue to stop the service will mark the runtime as stopped and allow the agent to create and restart the service runtime, and report successful or failed/missed check-in in the status, which  I think is the better experience, especially in the cases where the agent restarts the endpoint due to the logging level change for example. If we fail to stop the service runtime is left stuck in the failed state. 

## Why is it important?

Implements service runtime component.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas

## How to test this PR locally

1. Build the agent package.
2. Unpack the agent.
3. Copy over the endpoint binaries and the spec into the agent ```components``` directory
Screen Shot 2022-09-05 at 12 51 55 PM
4. Enroll and start the agent

## Related issues

- Closes https://github.com/elastic/elastic-agent/issues/1069

## Screenshots

Screen Shot 2022-10-13 at 9 37 02 AM

### CODE REVIEW

**Comment by aleksmaus** [comment]
Based on the today's conversation with the Endpoint there was a misunderstanding what the Agent suppose to do in terms of managing the service lifecycle. The way it is implemented here is that the agent does the best effort to start and stop the service across the platforms. In reality the agent should not do any of this.

The following changes will need to be made:
1. The agent should only check if the service is installed, install and uninstall the service
2. These check/install/uninstall commands should be noop in k8s environment
3. The agent should issue STOPPING signal to let endpoint handle stopping for K8s environment. The Endpoint "reverts to inert state based on the config removed".
4. Remove all the unnecessary code that evolves around platform services management APIs.

**Comment by aleksmaus** [comment]
Addressed the changes described in the previous comment.

**Comment by aleksmaus** [comment]
Changed the stopping/teardown logic a bit based on the latest conversation with Endpoint team. 
Now when the agent tears down/uninstalls endpoint (endpoint is removed from the policy):
1. The agent waits for up to 30 secs if the endpoint haven’t checked in
2. The agent sends STOPPING if the endpoint checked in
3. The agent calls uninstall on the endpoint binary

File: internal/pkg/agent/application/application.go [hunk 1]
```diff
@@ -159,6 +161,11 @@ func mergeFleetConfig(rawConfig *config.Config) (storage.Store, *configuration.C
 			errors.M(errors.MetaKeyPath, path))
 	}
 
+	// Fix up fleet.agent.id otherwise the fleet.agent.id is empty string
```
Comment (cmacknz) [inline internal/pkg/agent/application/application.go:164]: Why is this necessary?

File: internal/pkg/agent/application/fleet_server_bootstrap.go [hunk 2]
```diff
@@ -82,6 +83,37 @@ func FleetServerComponentModifier(serverCfg *configuration.FleetServerConfig) co
 	}
 }
 
+func EndpointComponentModifier(fleetCfg *configuration.FleetAgentConfig) coordinator.ComponentsModifier {
```
Comment (cmacknz) [inline internal/pkg/agent/application/fleet_server_bootstrap.go:86]: Can you document what this does, and why it is needed (i.e Endpoint currently needs to talk directly to Fleet server to upload files)?

File: pkg/component/runtime/conn_info_server.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+)
+
+type connInfoServer struct {
+	log         *logger.Logger
+	listener    net.Listener
+	waitCtx     context.Context
+	stopTimeout time.Duration
+}
+
+func newConnInfoServer(log *logger.Logger, comm Communicator, port int) (*connInfoServer, error) {
```
Comment (cmacknz) [inline pkg/component/runtime/conn_info_server.go:30]: Just reading this in isolation from its use cases its hard to know what problem this is solving.

Finding the definition of `WriteConnInfo` helped a little bit, in that I can see we are serialization a protobuf connection information message and writing it the writer: https://github.com/elastic/elastic-agent/blob/77beb04ffb47fd4b06860844fbdb7ef803276ca8/pkg/component/runtime/runtime_comm.go#L112-L119

It's not obvious who should be dialing us on localhost to obtain this information, or why.

I assume this is a special method of boostrapping the control protocol connection for components that run as independent services? Can we document this somewhere?

Could you create need a service runtime README.md file or package level documentation comment explaining how this works, and how it is all wired together? I think that would help people unfamiliar with this part of the agent ramp up on it faster.

File: pkg/component/spec.go [hunk 4]
```diff
@@ -81,10 +81,24 @@ func (t *CommandTimeoutSpec) InitDefaults() {
...
+
+// InitDefaults initialized the defaults for the timeouts.
+func (t *ServiceTimeoutSpec) InitDefaults() {
+	t.Checkin = 30 * time.Second
+	t.Stop = 3 * time.Minute
+}
+
 // ServiceSpec is the specification for an input that executes as a service.
 type ServiceSpec struct {
+	CPort      int                   `config:"cport" yaml:"cport" validate:"required"`
```
Comment (cmacknz) [inline pkg/component/spec.go:98]: What is CPort? Control protocol port?

File: pkg/component/runtime/conn_info_server.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	go func() {
+		defer cn()
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				log.Errorf("failed accept conn info connection: %v", err)
+				break
+			}
+			log.Debugf("client connected, sending connection info")
+			err = comm.WriteConnInfo(conn)
```
Comment (cmacknz) [inline pkg/component/runtime/conn_info_server.go:49]: Should we be doing any kind of verification of whose connection we accepted before handing out connection information?

**Review by cmacknz — COMMENTED** [review]
Did a first pass review, there is a lot to digest in here.

I think the biggest thing that would help us review and maintain this would be to write a README (or equivalent) explaining the design of how we bootstrap and manage service components. 

It may also be worth mentioning that services support is highly specialized to endpoint security, since that is currently the only service component we need to run.

File: internal/pkg/agent/application/application.go (hunk 1 above)
Comment (aleksmaus) [reply internal/pkg/agent/application/application.go:164]: Was working off of the endpoint fleet config example and the ```fleet.agent.id``` was empty from the fleet configuration file. The agent.id was at top level but not under the ```fleet``` which we suppose to pass to the Endpoint.

```
fleet:
  agent:
    id: 01010101-0101-0101-0101-010101010101
    logging:
      level: debug
  host:
    id: dabadaba-d000-0000-0000-000000000000
  access_api_key: akjdgskjhdgskhgdskhfgshjdfgsjhdgdfsjhdgfsjsdfjhgsdfagjshUQ==
  protocol: https
  hosts:
  - localhost:5601
  path: /boo
  proxy_disable: false
  proxy_url: ""
  ssl:
    verification_mode: full
    ca_trusted_fingerprint: 96BCEC06264976F37460779ACF28C5A7CFE8A3C0AAE11A8FFCEE05C0BDDF08C6
    certificate_authorities:
    - /tmp/dan/ca_certs
    certificate: "/etc/pki/client/cert.pem"
    key: "/etc/pki/client/cert.key"
    ca_sha256: 7HIpactkIAq2Y49orFOOQKurWxmmSFZhBCoQYcRhJ3Y=
```

File: internal/pkg/agent/application/fleet_server_bootstrap.go (hunk 2 above)
Comment (aleksmaus) [reply internal/pkg/agent/application/fleet_server_bootstrap.go:89]: Added comments

File: pkg/component/runtime/conn_info_server.go (hunk 3 above)
Comment (aleksmaus) [reply pkg/component/runtime/conn_info_server.go:30]: It was not documented in V1, so had to grep around to figure out how it works.
Sure can create a brief readme at the package level.

File: pkg/component/runtime/conn_info_server.go (hunk 5 above)
Comment (aleksmaus) [reply pkg/component/runtime/conn_info_server.go:49]: there is no additional verification in V1. This code implements exactly what V1 was doing, just refactored into a better encapsulated piece.

File: pkg/component/spec.go (hunk 4 above)
Comment (aleksmaus) [reply pkg/component/spec.go:98]: ```creds port``` or ```connection info port```, didn't want it to be confused with any other possible primary connection parameters between the agent and the endpoint

**Review by aleksmaus — COMMENTED** [review]
Addressed the code review comments

**Comment by aleksmaus** [comment]
```make check-ci``` passes locally and fails on CI. is this known issue? any workarounds? @cmacknz?
 
 ```
 ➜  elastic-agent git:(feature/service_runtime_v2) make check-ci
>> Building elastic-agent.yml for linux/amd64
>> Building elastic-agent.reference.yml for linux/amd64
>> Building elastic-agent.docker.yml for linux/amd64
>> BuildPGP from GPG-KEY-elasticsearch to internal/pkg/release/pgp.go
>> BuildFleetCfg _meta/elastic-agent.fleet.yml to internal/pkg/agent/application/configuration_embed.go
Generating NOTICE
go mod tidy
go mod download
go list -m -json all | go run go.elastic.co/go-licence-detector \
		-includeIndirect \
		-rules dev-tools/notice/rules.json \
		-overrides dev-tools/notice/overrides.json \
		-noticeTemplate dev-tools/notice/NOTICE.txt.tmpl \
		-noticeOut NOTICE.txt \
		-depsOut ""
cat dev-tools/notice/NOTICE.txt.append >> NOTICE.txt
Generating elastic-agent-standalone-kubernetes.yaml
Generating elastic-agent-managed-kubernetes.yaml
➜  elastic-agent git:(feature/service_runtime_v2) git status
On branch feature/service_runtime_v2
Your branch is up to date with 'origin/feature/service_runtime_v2'.

nothing to commit, working tree clean
```

File: pkg/component/runtime/conn_info_server.go (hunk 3 above)
Comment (cmacknz) [reply pkg/component/runtime/conn_info_server.go:30]: README is great, thank you!

File: pkg/component/runtime/conn_info_server.go (hunk 5 above)
Comment (cmacknz) [reply pkg/component/runtime/conn_info_server.go:49]: Thanks, it may be worth coming back to this in the future but we can leave it as is since this is the way it currently works.

File: specs/endpoint-security.spec.yml [hunk 6]
```diff
@@ -60,6 +62,7 @@ inputs:
         - condition: ${runtime.user.root} == false
           message: "Elastic Agent must be running as Administrator or SYSTEM"
     service:
+      cport: 6788
```
Comment (michalpristas) [inline specs/endpoint-security.spec.yml:65]: please include increased timeout from PR #1286

Comment (aleksmaus) [reply specs/endpoint-security.spec.yml:65]: Removed the timeout for Stop from V2.
Based on conversation with Endpoint team earlier. The STOPPING state issuing should be the best effort attempt and followed with uninstall. STOPPING is effectively NOOP for the normal Endpoint installation, since the uninstall will take care of stopping and uninstalling the Endpoint service. It's only needed for k8s installation support, so the endpoint can be "deactivated" in k8s environment.

Now when the Endpoint removed from the policy the following sequence happening:
1. If Endpoint has not checked-in yet, the Agent awaits for check-in up to the check-in timeout duration
2. The Agent sends STOPPING as expected state to the Endpoint if Endpoint checked-in
3. The Agent uninstalls Endpoint, logs the error if there was an error.
4. The Agent sets the service runtime into STOPPED state always. This avoids leaving the Agent in unrecoverable failed state.

File: pkg/component/runtime/runtime.go [hunk 7]
```diff
@@ -54,13 +54,13 @@ type ComponentRuntime interface {
 }
 
 // NewComponentRuntime creates the proper runtime based on the input specification for the component.
-func NewComponentRuntime(comp component.Component, monitor MonitoringManager) (ComponentRuntime, error) {
+func NewComponentRuntime(comp component.Component, monitor MonitoringManager, logger *logger.Logger) (ComponentRuntime, error) {
```
Comment (michalpristas) [inline pkg/component/runtime/runtime.go:57]: convention

```suggestion
func NewComponentRuntime(comp component.Component, logger *logger.Logger, monitor MonitoringManager)
```

File: pkg/component/runtime/runtime.go [hunk 8]
```diff
@@ -87,7 +87,7 @@ func newComponentRuntimeState(m *Manager, logger *logger.Logger, monitor Monitor
 	if err != nil {
 		return nil, err
 	}
-	runtime, err := NewComponentRuntime(comp, monitor)
+	runtime, err := NewComponentRuntime(comp, monitor, logger)
```
Comment (michalpristas) [inline pkg/component/runtime/runtime.go:90]: related to L57
```suggestion
        runtime, err := NewComponentRuntime(comp, logger, monitor)
```

File: pkg/component/runtime/service_command.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	opts := []process.StartOption{
+		process.WithContext(ctx),
+		process.WithArgs(args),
+		process.WithEnv(env),
+	}
+
+	// Set the command working directory from binary
+	// This is needed because the endpoint installer was looking for it's resources in the current working directory
+	wdir := filepath.Dir(binaryPath)
+	if wdir != "." {
```
Comment (michalpristas) [inline pkg/component/runtime/service_command.go:42]: this should never happen (wdir to be `.`) we cannot use relative paths, but i dont mind having it here.

File: pkg/component/runtime/service_command.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	proc, err := process.Start(binaryPath, opts...)
+	if err != nil {
+		return fmt.Errorf("failed starting the command: %w", err)
+	}
+
+	// channel for the last error message from the stderr output
+	errch := make(chan string, 1)
+	ctxstderr := contextio.NewReader(ctx, proc.Stderr)
```
Comment (michalpristas) [inline pkg/component/runtime/service_command.go:57]: naming: please use something like `ctxStdErr` or `ctxStderr`

File: pkg/component/runtime/service_command.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+	if ctxstderr != nil {
+		go func() {
+			var errtext string
```
Comment (michalpristas) [inline pkg/component/runtime/service_command.go:60]: s/errtext/errText

File: pkg/component/runtime/service_command.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+			scanner := bufio.NewScanner(ctxstderr)
+			for scanner.Scan() {
+				line := scanner.Bytes()
```
Comment (michalpristas) [inline pkg/component/runtime/service_command.go:63]: you can use `scanner.Text()` you convert it 2 times anyways

File: pkg/component/runtime/service_command.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+			err = fmt.Errorf("%s: %w", errmsg, err)
+		}
+	}
+
+	return err
+}
+
+func executeServiceCommand(ctx context.Context, log *logger.Logger, binaryPath string, spec *component.ServiceOperationsCommandSpec) error {
+	if spec == nil {
+		return nil
```
Comment (michalpristas) [inline pkg/component/runtime/service_command.go:97]: is this expected? should we log warning?

File: pkg/component/runtime/state.go [hunk 14]
```diff
@@ -400,6 +400,21 @@ func (s *ComponentState) forceState(state client.UnitState, msg string) bool {
 	return changed
 }
 
+// forceExpectedState force updates the expected state for the entire component, forcing that state on all expected units.
+func (s *ComponentState) forceExpectedState(state client.UnitState) bool {
```
Comment (michalpristas) [inline pkg/component/runtime/state.go:404]: not using return value anywhere

File: pkg/component/spec.go [hunk 15]
```diff
@@ -81,10 +81,22 @@ func (t *CommandTimeoutSpec) InitDefaults() {
...
 }
 
+// ServiceTimeoutSpec is the timeout specification for subprocess.
+type ServiceTimeoutSpec struct {
+	Checkin time.Duration `config:"checkin" yaml:"checkin"`
+}
+
+// InitDefaults initialized the defaults for the timeouts.
+func (t *ServiceTimeoutSpec) InitDefaults() {
+	t.Checkin = 30 * time.Second
```
Comment (michalpristas) [inline pkg/component/spec.go:91]: didn't you say 3 minutes?

**Review by michalpristas — COMMENTED** [review]
ping me for round 2

File: pkg/component/spec.go (hunk 15 above)
Comment (aleksmaus) [reply pkg/component/spec.go:91]: check-in is always 30 secs default.
see my last comment on this PR:
https://github.com/elastic/elastic-agent/pull/1529#discussion_r1004432188

File: go.mod [hunk 16]
```diff
@@ -152,6 +153,7 @@ require (
 
 replace (
 	github.com/Microsoft/go-winio => github.com/bi-zone/go-winio v0.4.15
+	github.com/Shopify/sarama => github.com/elastic/sarama v1.19.1-0.20220310193331-ebc2b0d8eef3
```
[the remark this answers is not in the document]
Comment (aleksmaus) [reply go.mod:156]:
```
➜  elastic-agent git:(feature/service_runtime_v2) rm -rf ./build
➜  elastic-agent git:(feature/service_runtime_v2) go mod tidy
➜  elastic-agent git:(feature/service_runtime_v2) git status
On branch feature/service_runtime_v2
Your branch is up to date with 'origin/feature/service_runtime_v2'.

nothing to commit, working tree clean
```

File: internal/pkg/agent/application/fleet_server_bootstrap.go [hunk 17]
```diff
@@ -82,6 +83,40 @@ func FleetServerComponentModifier(serverCfg *configuration.FleetServerConfig) co
...
+						unitCfg, err := component.ExpectedConfig(unitCfgMap)
+						if err != nil {
+							return nil, err
+						}
+						unit.Config = unitCfg
+					}
+					comp.Units[j] = unit
+				}
+			}
+			comps[i] = comp
```
[the remark this answers is not in the document]
Comment (aleksmaus) [reply internal/pkg/agent/application/fleet_server_bootstrap.go:114]: consistent with existing FleetServerComponentModifier implementation

**Review by michalpristas — APPROVED** [review]
lgtm
